src/sig/sig.h

```diff
@@ -63,108 +84,123 @@ struct OQS_SIG {
 	size_t length_public_key;
 	/** The (maximum) length, in bytes, of secret keys for this signature scheme. */
 	size_t length_secret_key;
-	/** The (maximum) length, in bytes, of a signature for this signature scheme. */
-	size_t max_length_signature;
+	/** The (maximum) length, in bytes, of signatures for this signature scheme. */
+	size_t length_signature;
 
-	/**
-	 * Opaque pointer for passing around any computation context
-	 */
-	void *ctx;
-
-	// clang-format and doxygen don't work together nicely on these three functions
 	// clang-format off
+
 	/**
-	 * Pointer to a function for public and private signature key generation.
+	 * Keypair generation algorithm.
+	 *
+	 * Caller is responsible for allocating sufficient memory for `public_key` and
+	 * `secret_key`, based on the `length_*` members in this object or the per-scheme
+	 * compile-time macros `OQS_SIG_*_length_*`.
 	 *
-	 * @param s                The signature structure.
-	 * @param priv             The signer's private key.
-	 * @param pub              The signer's public key.
-	 * @return                 OQS_SUCCESS on success, or OQS_ERROR on failure.
+	 * @param[out] public_key The public key represented as a byte string.
+	 * @param[out] secret_key The secret key represented as a byte string.
+	 * @return OQS_SUCCESS or OQS_ERROR
 	 */
-	OQS_STATUS(*keygen)(const OQS_SIG *s, uint8_t *priv, uint8_t *pub);
+	OQS_STATUS (*keypair)(uint8_t *public_key, uint8_t *secret_key);
 
 	/**
-	 * Pointer to a function for signature generation.
+	 * Signature generation algorithm.
 	 *
-	 * @param s                The signature structure.
-	 * @param priv             The signer's private key.
-	 * @param msg              The message to sign.
-	 * @param msg_len          Length of the message to sign.
-	 * @param sig              The generated signature. Must be allocated by the caller, or NULL to learn how much space is needed, as returned in sig_len.
-	 * @param sig_len          In: length of sig, out: length of the generated signature.
-	 * @return                 OQS_SUCCESS on success, or OQS_ERROR on failure.
+	 * Caller is responsible for allocating sufficient memory for `signed_message`,
+	 * based on the `length_*` members in this object or the per-scheme
+	 * compile-time macros `OQS_SIG_*_length_*`.
+	 *
+	 * @param[out] signature The signature on the message represented as a byte string.
+	 * @param[out] signature_len The length of the signature.
+	 * @param[in] message The message to sign represented as a byte string.
+	 * @param[in] message_len The length of the message to sign.
+	 * @param[in] secret_key The secret key represented as a byte string.
+	 * @return OQS_SUCCESS or OQS_ERROR
 	 */
-	OQS_STATUS(*sign)(const OQS_SIG *s, const uint8_t *priv, const uint8_t *msg, const size_t msg_len, uint8_t *sig, size_t *sig_len);
+	OQS_STATUS (*sign)(uint8_t *signature, size_t *signature_len, const uint8_t *message, size_t message_len, const uint8_t *secret_key);
```

Comment:
`message_len` should be marked as const

Reply:
Yes; leftovers from copy-pasting from nist-branch

Reply:
Well we do it for uint8_t which can't be changed by the function either, so why one and not the other?

Reply:
we do it for uint8_t _pointers_, the const being a promise to the caller that the function won't modify the content. I don't see why that would be useful for a size parameter.